### ISSUE TITLE
Interesting work

### ISSUE THREAD

**Issue by zwzhang123** [ahmdtaha/svmax#1]
I used to devel into the researches on singular value of feature embeddings. 

If you maximize the mean of  singular values, how about the training curves max & min?

**Comment by zwzhang123** [comment]
How do you think the role of singular values' distribution on image classification?
Is uniform feature distribution more discrimitive?

**Comment by ahmdtaha** [comment]
Thanks for your interest.
Q1. If you maximize the mean of singular values, how about the training curves max & min?
If the feature embedding is l2-normalized, the max and min singular values will decrease and increase respectively, as shown in Fig.1
If the feature embedding is not l2-normalized, the max and min depend on the hyperparameter \lambda. 
Let's assume L = attrat -\lambda s_\mu. If \lambda is small, the first term (attraction term) will take over and a model collapse will happen -- all singular values go to zero.
If \lambda is big, SVMax dominates the first term. The network embedding will go to infinity -- all singular values will increase.
Basically, you need to carefully tune \lambda to avoid any extreme solution.

**Comment by ahmdtaha** [comment]
Q2. How do you think the role of singular values' distribution on image classification? Is uniform feature distribution more discrimitive?

It is difficult to answer this question because cross-entropy makes no explicit assumption about the feature embedding. However, if you take cross-entropy regularizers (Center-loss [1] and AdaCos [2]) into consideration, then a classification network will learn an embedding that promotes class compactness and margin maximization (as in Fig-1.a). This discriminative feature embedding is good for classification networks because the testing classes are the same as training classes -- closed-set recognition. In Fig-1.a, you have two classes, you want to maximize the margin between them -- good for classification. 

However, for retrieval networks, the testing classes are different from the training classes. You want to cover the whole feature embedding to handle unseen classes -- open-set recognition.

For a typical classification network, I would say the feature embedding should be less uniform.


[1] A discriminative feature learning approach for deep face recognition
[2] AdaCos: Adaptively Scaling Cosine Logits for Effectively Learning Deep Face Representations

**Comment by zwzhang123** [comment]
Thank you for your detailed answers. I will review the above two papers.

oh, for Q2, I actually want to ask “Is uniform distribution of singular values more discrimitive?”

I remember paper [1] claims that "the eigenvectors with the largest singular values will dominate the feature transferability. As a consequence, the transferability is enhanced at the expense of over penalization of other eigenvectors that embody rich structures crucial for discriminability. Towards this problem, we present Batch Spectral Penalization (BSP), a general approach to **penalizing the largest singular values** so that other eigenvectors can be relatively strengthened to boost the feature discriminability".

Domain adaptation is an open-set recognition task. It seems that the answer of Q2 is YES! Do you think so?
How about for closed-set recognition?

**Comment by zwzhang123** [comment]
In addition, I think the direction of eigenvectors may be a important point. But I have not figured it out yet!
how do you think it?

**Comment by ahmdtaha** [comment]
What is [1] in your reply? Are you referring to center-loss or a different paper?

**Comment by zwzhang123** [comment]
oh, I forgot add the paper's name!!

[1] Transferability vs. Discriminability: Batch Spectral Penalization for Adversarial Domain Adaptation

**Comment by ahmdtaha** [comment]
Q2. Is uniform distribution of singular values more discriminative?
In open-set recognition, short answer: Yes. Long answer: Yes, but this better be a **prior** in empirical experiments. In the same dataset, _different_ classes are not uniformly different,i.e., some classes are similar to each other compared to other classes. 

In closed-set recognition, I am really not sure. I will explain my confusion using both SVMax and BSP terminology
[SVMax Terminology] On one hand, you want different classes to spread away from each other in the feature embedding (high inter-class). Yet, you also want class compactness -- do not spread that much (low intra-class). These are two contradicting objectives from a feature embedding point of view.

[BSP Terminology] In a closed-set, if the eigenvector with the largest singular values is enough to deliver discriminability, then you are good to go. For instance, imagine if you project a 10-class dataset on a 1D axis _equally spaced_ -- between 0-10, this would not be a bad solution. You will see the same classes during testing and you only need one dimension to differentiate between them. Not bad at all.


I just want to highlight one difference between SVMax and BSP. SVMax maximizes the mean singular value, while BSP minimizes the maximum singular values. On a high level, these two objectives are similar. However, without both the L2-normalization and the singular bounds analysis in SVMax, BSP is sensitive to its hyperparameter. Concretely, the max singular value can be 1 or 100 depending on factors like the batch-size, embedding-dimension, and the values inside the embedding matrix. Tuning hyperparameter is not a hurdle for some people. But I don't like this process. It feels like a waste of time.

I have no idea whether the direction of eigenvectors makes a difference or not. Can't help with this one; sorry.

**Comment by zwzhang123** [comment]
Thank you very much!
I understand that distribution of singular values should be uniform in the open-set recognition. 
There is some uncertainty? Maybe the best way is to show them by figures. Conclusions should be based on statistical results, I want to check it in free time.

For the direction of eigenvectors, I need to think about it more carefully.

**Comment by ahmdtaha** [comment]
I have one final comment to avoid any misunderstanding. I will quote the SVMax paper 

> Weight decay imposes a prior on the network weights. Similarly, SVMax imposes a prior on the network activations. Both priors are important but they should never dominate the loss function.
